### PR TITLE
Migration from crossref to ROR

### PR DESCRIPTION
**Pull request by MSDrao**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- 

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [ ] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### CODE REVIEW

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 1]
```diff
@@ -36,10 +36,12 @@ let fundingAgenciesApp = new Vue({
     crossreffApiDown: false, // if we are having trouble reaching the crossref api
     removeCharsFromQuery: ["."], // characters to be removed from search
     filteredWords: [],
+    ROR_API_URL: "https://api.ror.org/v2/organizations"
   },
   mounted() {
-    if (this.resourceMode === 'Edit' && (this.selfAccessLevel === 'owner' || this.selfAccessLevel === 'editor') && this.resPublished) {
-      this.checkFunderNamesExistInCrossref(this.fundingAgencies);
+    if (this.resourceMode === 'Edit' && (this.selfAccessLevel === 'owner' || this.selfAccessLevel === 'editor') ) {
```
Comment (devincowan) [inline theme/static/js/hs-vue/funding-agency-app.js:42]: why did you remove the condition for `this.resPublished` here?

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 2]
```diff
@@ -36,10 +36,12 @@ let fundingAgenciesApp = new Vue({
...
+      // this.checkFunderNamesExistInCrossref(this.fundingAgencies);
```
Comment (devincowan) [inline theme/static/js/hs-vue/funding-agency-app.js:43]: remove this commented out line
and we can get rid of the `checkFunderNamesExistInCrossref()` function

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 3]
```diff
@@ -120,12 +155,57 @@ let fundingAgenciesApp = new Vue({
...
+      return null;
+    },
+
     initCrossrefFundersSearch: async function (funderName) {
       if (funderName === "" || this.notifications.error) {
         return;
       }
       this.isPending = true;
-      this.crossrefFunders = await this.fetchFromCrossrefAPIFunderList(
+      this.crossrefFunders = await this.fetchFromRORList(
```
Comment (devincowan) [inline theme/static/js/hs-vue/funding-agency-app.js:208]: rename `this.crossrefFunders` to avoid confusion

File: hs_core/models.py [hunk 4]
```diff
@@ -2007,6 +2007,47 @@ def update(cls, element_id, **kwargs):
 
         super(FundingAgency, cls).update(element_id, **kwargs)
 
+    @classmethod
+    def migrate_from_crossref_ror(cls, apps, schema_editor):
+        """Migrate from CrossRef to ROR"""
+        logger = logging.getLogger(__name__)
+
+        for page in range(1, 500):
```
Comment (devincowan) [inline hs_core/models.py:2015]: why limit to 500?

File: hs_core/models.py [hunk 5]
```diff
@@ -2007,6 +2007,47 @@ def update(cls, element_id, **kwargs):
 
         super(FundingAgency, cls).update(element_id, **kwargs)
 
+    @classmethod
+    def migrate_from_crossref_ror(cls, apps, schema_editor):
+        """Migrate from CrossRef to ROR"""
```
Comment (devincowan) [inline hs_core/models.py:2012]: if it would be helpful, I can get you a db dump with the exsiting FundingAgencies so that you can test this migration?

File: hs_core/models.py [hunk 6]
```diff
@@ -2007,6 +2007,47 @@ def update(cls, element_id, **kwargs):
...
+                logger.debug('Iterating through ROR items.')
+                for item in items:
+                    logger.debug(f"ROR ID : {item['id']}")
+                    if item['types'] and 'funder' in item['types']:
+                        external_ids = item['external_ids']
+                        if external_ids:
+                            fundrefs = [x_id for x_id in external_ids if x_id['type'] == 'fundref']
+                            if len(fundrefs) == 0:
+                                continue
+                            fundref = fundrefs[0]
```
Comment (devincowan) [inline hs_core/models.py:2033]: do we need to check more than just the first `fundref[0]`?

File: hs_core/models.py [hunk 7]
```diff
@@ -2007,6 +2007,47 @@ def update(cls, element_id, **kwargs):
...
+                            agency_url_prefix = 'http://dx.doi.org/10.13039/'
+                            agency_url = agency_url_prefix + fundref_id
+
+                            funding_agency_model = apps.get_model('hs_core', 'FundingAgency')
+                            agency = funding_agency_model.objects.filter(agency_url=agency_url).first()
+                            agency_name_new = [n['value'] for n in item['names'] if 'ror_display' in n['types']][0]
+                            if agency is not None:
+                                logger.debug(f'Updating funding agency with {agency_url}, {agency_name_new}')
+                                funding_agency_model.objects.filter(id=agency.id).update(agency_url=item['id'],
+                                                                                         agency_name=agency_name_new)
```
Comment (devincowan) [inline hs_core/models.py:2049]: when you do `funding_agency_model.objects.filter(agency_url=agency_url).first()` you are taking only the first matching FundingAgency. Could there be multiple matches that need to be updated?

File: hs_core/models.py [hunk 8]
```diff
@@ -2007,6 +2007,47 @@ def update(cls, element_id, **kwargs):
...
+            url = f"https://api.ror.org/v2/organizations?page={page}&filter=types:funder"
+            response = requests.get(url, verify=False)
+            if response.status_code == 200:
+                response_json = response.json()
+                logger.debug(f"Number of items: {response_json.get('number_of_results', None)}")
+                items = response_json.get('items', None)
+                if not items:
+                    break
+                logger.debug('Iterating through ROR items.')
+                for item in items:
```
Comment (devincowan) [inline hs_core/models.py:2025]: this migration will take a long time (5min?) EVEN IF WE ARE RUNNING IT ON A DB THAT HAS NO FUNDING AGENCIES! :)

Instead of getting all of the funding agencies from ROR and iterating through them, I wonder if it would be faster to iterate over all of the Funding Agencies that exist in our HydroShare db, and for each of those fundingAgencies, making a request to the ROR API...?

Let's see. We have ~5000 funding agencies in the HS db. Some of them have null agency_url
```py
>>> FundingAgency.objects.exclude(agency_url__isnull=True).count()
4558
```
Okay so maybe your approach is faster than making ~4558 synchronous requests. But we should consider making this asynchronous.

Here is an example task that we could use to ponder:
https://github.com/hydroshare/hydroshare/blob/2.15.8/hs_core/tasks.py#L434-L436

File: theme/templates/resource-landing-page/funding-agency.html [hunk 9]
```diff
@@ -74,7 +74,7 @@ <h4 class="modal-title" id="add-funding-label"> ${mode} Funding Agency</h4></div
 
                                     <div class="controls">
                                         <input class="form-control input-sm urlinput" maxlength="200"
-                                            name="agency_url" type="url" v-model="currentlyEditing.agency_url" @keyup="checkAgencyName">
+                                            name="agency_url" type="url" v-model="currentlyEditing.agency_url" @keyup="checkAgencyName" readonly>
```
Comment (devincowan) [inline theme/templates/resource-landing-page/funding-agency.html:77]: nice check on this `readonly`
I agree that it seems like we wouldn't want the user to be able to edit the agency_url.
But we actually want to allow editing the url. This was discussed in initial implementation and decided best to allow the user additional flexibility

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 10]
```diff
@@ -120,12 +155,57 @@ let fundingAgenciesApp = new Vue({
...
+        this.crossreffApiDown = false;
+        return funders;
+      } catch (e) {
+        console.error(`Error querying Crossref API: ${e}`);
+        this.crossreffApiDown = true;
+      }
+      return null;
+    },
+
     initCrossrefFundersSearch: async function (funderName) {
```
Comment (devincowan) [inline theme/static/js/hs-vue/funding-agency-app.js:203]: also rename this function, to make it clear that we are no longer searching the Crossref funder list

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 11]
```diff
@@ -120,12 +155,57 @@ let fundingAgenciesApp = new Vue({
       }
       return null;
     },
+    fetchFromRORList: async function (funderName) {
```
Comment (devincowan) [inline theme/static/js/hs-vue/funding-agency-app.js:158]: Nice! This looks good to me @MSDrao!

**Review by devincowan — CHANGES_REQUESTED** [review]
Overall this looks really great @MSDrao! Some questions for you to consider...

File: hs_core/models.py (hunk 8 above)
Comment (MSDrao) [reply hs_core/models.py]: After reviewing both options, I agree that many synchronous API calls would slow things down. I'll make it async and this speeds things up without blocking other tasks.

File: hs_core/models.py (hunk 7 above)
Comment (MSDrao) [reply hs_core/models.py]: Good catch! I’ll adjust the code to update all matching records instead of just the first one. This will make sure everything stays consistent.

File: theme/static/js/hs-vue/funding-agency-app.js (hunk 1 above)
Comment (MSDrao) [reply theme/static/js/hs-vue/funding-agency-app.js]: I removed the this.resPublished condition because I wanted to allow funder name checks even before the resource is published. This helps catch any issues earlier. But, after doing some tests in Beta version, I thought it's better to keep it the this.resPublished condition.

**Comment by devincowan** [comment]
@MSDrao please re-request review once your done making changes. Thanks!

**Comment by devincowan** [comment]
@MSDrao this is looking pretty good, thanks for this work!
Let's run your migration on beta sometime this week and once that box is checked, I will approve.

**Review by devincowan — CHANGES_REQUESTED** [review]

File: theme/static/js/hs-vue/funding-agency-app.js [hunk 12]
```diff
@@ -12,9 +12,9 @@ let fundingAgenciesApp = new Vue({
     resourceMode: RESOURCE_MODE, // edit/view
     selfAccessLevel: SELF_ACCESS_LEVEL, // user's access level on resource
     resPublished: RESOURCE_PUBLISHED_OR_UNDER_REVIEW,
-    crossrefFunders: [], // array of funders to be filled from crossref api
```
[the remark this answers is not in the document]
Comment (MSDrao) [reply theme/static/js/hs-vue/funding-agency-app.js:15]: Fixed this issue @devincowan

File: hs_core/models.py (hunk 4 above)
Comment (devincowan) [reply hs_core/models.py]: makes sense thanks

File: theme/static/js/hs-vue/funding-agency-app.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (devincowan) [reply theme/static/js/hs-vue/funding-agency-app.js:15]: thanks, looks good

**Review by devincowan — APPROVED** [review]
Nice work @MSDrao!
After you verify that everything looks good on beta, you can merge this PR

**Review by devincowan — CHANGES_REQUESTED** [review]
@MSDrao I realized that there is another issue
When we submit a resource for publication, we include the funding_agency metadata in the publication request.

So this PR will need also to update all of the publication metadata

Here is where the funding metadata are added into the Crosreff XML:
https://github.com/hydroshare/hydroshare/blob/5453953c34f290443aecd11d7b7ff90a5ecef908/hs_core/models.py#L4067-L4087

File: hs_core/models.py [hunk 13]
```diff
@@ -2035,6 +2036,7 @@ async def update_record(self, record):
                         record.agency_name = agency_name_new
                         record.agency_url = item['id']
                         await sync_to_async(record.save)()
+                        await update_crossref_meta_deposit.apply_async((self.short_id,))
```
Comment (devincowan) [inline hs_core/models.py:2039]: `FundingAgency` does not have a `short_id`

**Review by devincowan — CHANGES_REQUESTED** [review]

File: hs_core/models.py (hunk 13 above)
Comment (devincowan) [reply hs_core/models.py:2039]: adding more explanation as requested here @MSDrao 
[update_crossref_meta_deposit](https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/tasks.py#L1245) takes a resource id as an argument.

You are attempting to call `await update_crossref_meta_deposit.apply_async((self.short_id,))` here:
https://github.com/hydroshare/hydroshare/blob/crosref_to_ror_migrate_develop/hs_core/models.py#L2039
from within the [FundingAgency](https://github.com/hydroshare/hydroshare/blob/crosref_to_ror_migrate_develop/hs_core/models.py#L1981) class. In this context, `self.short_id` does not resolve to a resource id. You are trying to get the [AbstractResource.short_id](https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/models.py#L2148) for the resource associated with the given FundingAgency that is being updated in your [update_record](https://github.com/hydroshare/hydroshare/blob/crosref_to_ror_migrate_develop/hs_core/models.py#L2018) function

File: hs_core/models.py [hunk 14]
```diff
@@ -2007,6 +2009,46 @@ def update(cls, element_id, **kwargs):
 
         super(FundingAgency, cls).update(element_id, **kwargs)
 
+    @sync_to_async
+    def get_funding_records(self, apps):
+        FundingAgency = apps.get_model('hs_core', 'FundingAgency')
+        return list(FundingAgency.objects.all())
+
+    @classmethod
+    async def update_record(self, record):
```
Comment (devincowan) [inline hs_core/models.py:2018]: also a nomenclature recommendation:
this is a classmethod, so instead of `self` we typically use `cls` parameter

File: hs_core/models.py [hunk 15]
```diff
@@ -2007,6 +2009,46 @@ def update(cls, element_id, **kwargs):
...
+                        item = items[0]
+                        logger.debug(f"Replacing old funder id {record.agency_url} with new id {item['id']}")
+                        agency_name_new = [n['value'] for n in item['names'] if 'ror_display' in n['types']][0]
+                        record.agency_name = agency_name_new
+                        record.agency_url = item['id']
+                        await sync_to_async(record.save)()
+                        await update_crossref_meta_deposit.apply_async((self.short_id,))
+
+    @classmethod
+    async def get_funder_records(self, apps):
```
Comment (devincowan) [inline hs_core/models.py:2042]: classmethod should use `cls` instead of `self`

File: hs_core/models.py (hunk 13 above)
Comment (devincowan) [reply hs_core/models.py:2039]: `FundingAgencies` are linked to resources through CoreMetaData objects
https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/models.py#L4411

File: hs_core/models.py [hunk 16]
```diff
@@ -2036,18 +2036,22 @@ async def update_record(self, record):
                         record.agency_name = agency_name_new
                         record.agency_url = item['id']
                         await sync_to_async(record.save)()
-                        await update_crossref_meta_deposit.apply_async((self.short_id,))
+                        metadata = record.metadata
```
Comment (devincowan) [inline hs_core/models.py:2039]: you're headed in the right direction here!
But `record.metadata` is going to give you an error
Here your "record" is a FundingAgency object.
[FundingAgency](https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/models.py#L1979) class does not have any "metadata" attribute.

```
Running migrations:
  Applying hs_core.0086_auto_20241022_1602_crossref_to_ror_migration.../usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py:1063: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.ror.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py:1063: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.ror.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py:1063: InsecureRequestWarning: Unverified HTTPS request is being made to host 'api.ror.org'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
Traceback (most recent call last):
  File "/hydroshare/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/usr/local/lib/python3.9/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/hydroshare/hs_core/models.py", line 2054, in migrate_from_crossref_to_ror
    loop.run_until_complete(cls.get_funder_records(apps))
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/hydroshare/hs_core/models.py", line 2048, in get_funder_records
    await asyncio.gather(*(cls.update_record(record) for record in funding_records))
  File "/hydroshare/hs_core/models.py", line 2039, in update_record
    metadata = record.metadata
AttributeError: 'FundingAgency' object has no attribute 'metadata'
```

Comment (devincowan) [reply hs_core/models.py:2039]: Please test your migration against real data before requesting review again.
You can do that by creating a resource on `develop` branch and then checking out this PR branch and running `./hsctl managepy migrate`

**Review by devincowan — CHANGES_REQUESTED** [review]

File: hs_core/models.py (hunk 16 above)
Comment (devincowan) [reply hs_core/models.py:2039]: You can get the resource short_id for a given funding agency by using the [AbstractMetadataElement.content_object](https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/models.py#L450)

```py
from hs_core.models import FundingAgency
fa = FundingAgency.objects.first()
resource_id = fa.content_object.resource.short_id
```

File: hs_core/models.py (hunk 13 above)
Comment (devincowan) [reply hs_core/models.py]: Apologies! While true, [this last comment](https://github.com/hydroshare/hydroshare/pull/5632#discussion_r1842819155) ☝️ is misleading. The GenericRelation on CoreMetaData allows you to relate from BaseResource to FundingAgency, but for the converse relation you need the [AbstractMetadataElement.content_object](https://github.com/hydroshare/hydroshare/blob/2.15.9/hs_core/models.py#L450)

**Review by devincowan — APPROVED** [review]
Nice work @MSDrao!
